File: Models/FleetAvailability/fleet_avail_6seeds.py
#!/usr/bin/env python

from functools import partial
import numpy as np
from event_graph_sim import SimpleKit
from adaptive_stats import QuickStats
from argparse import ArgumentParser

# utility to produce 2-arg weibulls
def weib2(gen, alpha, beta):
    return gen(alpha) / beta


# Demonstration model of Fleet Availability

class FleetAvail(SimpleKit):

    def __init__(
        self,
        maxVehicles,
        maxMaintainers,
        breakdownRate,
        maintenanceCycleInDays,
        pStdRepair,
        pStdMaint,
        distributionChoice,
        alpha,
        beta,
        haltTime,
        outputType,
        rngs
    ):
        # The following line is mandatory for all SimpleKit models
        SimpleKit.__init__(self)  # DO NOT REMOVE!!!!
        # super().__init__()

        # model state
        self.dailyStats = QuickStats()
        self.results = []

        # model parameters
        self.maxVehicles = maxVehicles
        self.maxMaintainers = maxMaintainers
        self.breakdownRate = breakdownRate
        self.maintenanceCycleInDays = maintenanceCycleInDays
        self.pStdRepair = pStdRepair
        self.pStdMaint = pStdMaint
        self.repairRate = 1.0   # 1.0 / meanRepairTime
        self.distributionChoice = distributionChoice
        self.haltTime = haltTime
        self.outputType = outputType
        self.rngs = rngs

        self.numAvailableVehicles = self.maxVehicles
        self.numAvailableMechanics = self.maxMaintainers
        self.maintenanceQLength = 0
        self.breakdownQLength = 0

        match(outputType):
            case 's':
                labels = "initial stock,"
                labels += "#maintainers,"
                labels += "breakdown rate,"
                labels += "maintenance cycle,"
                # labels += "repair time,"
                labels += "p(std repair),"
                labels += "p(std maint),"
                # labels += "distribution,"
                labels += "alpha,"
                labels += "beta,"
                labels += "run length"
                labels += ",average availability,"
                labels += "stddev availability,"
                labels += "min,"
                labels += "10th percentile,"
                labels += "20th percentile,"
                labels += "median"
                print(labels)
                print(
                    f"{maxVehicles},{maxMaintainers},{breakdownRate},"
                    + f"{maintenanceCycleInDays},{pStdRepair},{pStdMaint},"
                    + f"{alpha},{beta},{haltTime / 8},",
                    end=""
                )
            case 'f':
                labels = "initial stock,"
                labels += "#maintainers,"
                labels += "breakdown rate,"
                labels += "maintenance cycle,"
                labels += "p(std repair),"
                labels += "p(std maint),"
                labels += "alpha,"
                labels += "beta,"
                labels += "run length"
                print(labels)
                print(
                    f"{maxVehicles},{maxMaintainers},{breakdownRate},"
                    + f"{maintenanceCycleInDays},{pStdRepair},{pStdMaint},"
                    + f"{alpha},{beta},{haltTime / 8}"
                )
                print("\n\nday,availability")

        # RV::Exponential.new(rate: self.breakdownRate, rng: rngs[0])
        self.breakdownDistribution = partial(rngs[0].exponential, scale=1.0/self.breakdownRate)

        # self.stdRepairDistribution = case self.distributionChoice
        match self.distributionChoice:
            case 'e':
                # self.stdRepairDistribution = RV::Exponential.new(rate: self.repairRate, rng: rngs[1])
                self.stdRepairDistribution = partial(rngs[1].exponential, scale=1.0/self.repairRate)
            case 'u':
                # self.stdRepairDistribution = RV::Uniform.new(min: 0, max: 2.0 / self.repairRate, rng: rngs[1])
                self.stdRepairDistribution = partial(rngs[1].uniform, low=0, high=2.0/self.repairRate)
            case 'w':
                # self.stdRepairDistribution = RV::Weibull.new(rate: alpha, k: beta, rng: rngs[1])
                self.stdRepairDistribution = partial(weib2, gen=rngs[1].weibull, alpha=alpha, beta=beta)

        match self.distributionChoice:
            case 'e':
                # self.brokenRepairDistribution = RV::Exponential.new(rate: self.repairRate / 4.0, rng: rngs[2])
                self.brokenRepairDistribution = partial(rngs[2].exponential, scale=4.0/self.repairRate)
            case 'u':
                # self.brokenRepairDistribution = RV::Uniform.new(min: 0, max: 8.0 / self.repairRate, rng: rngs[2])
                self.brokenRepairDistribution = partial(rngs[2].uniform, low=0, high=8.0/self.repairRate)
            case 'w':
                # self.brokenRepairDistribution = RV::Weibull.new(rate: 4.0 * alpha, k: beta, rng: rngs[2])
                self.brokenRepairDistribution = partial(weib2, gen=rngs[2].weibull, alpha=4.0*alpha, beta=beta)

        # self.stdMaintenanceDistribution = RV::Uniform.new(min: 5.5/8, max: 6.5/8, rng: rngs[3])
        self.stdMaintenanceDistribution = partial(rngs[3].uniform, low=5.5/8.0, high=6.5/8.0)


    # Event methods follow...

    def shutDown(self):
        self.results.sort()
        n = len(self.results) + 1
        if self.outputType == 's':
            print(
                f"{self.dailyStats.avg},{self.dailyStats.std_dev},"
                + f"{self.results[0]},{self.results[n // 10]},"
                + f"{self.results[n // 5]},{self.results[n // 2]}"
            )
        elif self.outputType == 'f':
            print(f"{int(self.model_time / 8.0)},{self.numAvailableVehicles}")
        self.halt()

    def dailyReport(self):
        if self.model_time < self.haltTime:
            self.results.append(self.numAvailableVehicles)
            self.dailyStats.new_obs(self.numAvailableVehicles)
            if self.outputType == 'f':
                print(f"{int(self.model_time / 8.0)},{self.numAvailableVehicles}")
        self.schedule(self.dailyReport, 8.0)

    def breakdown(self):
        self.breakdownQLength += 1
        self.numAvailableVehicles -= 1
        if (self.numAvailableMechanics > 0):
            self.schedule(self.beginBreakdownService, 0.0)

    def maintenance(self):
        self.maintenanceQLength += 1
        self.numAvailableVehicles -= 1
        if (self.numAvailableMechanics > 0):
            self.schedule(self.beginMaintenanceService, 0.0)

    def beginMaintenanceService(self):
        self.maintenanceQLength -= 1
        self.numAvailableMechanics -= 1
        if (self.rngs[4].random() <= self.pStdMaint):
            self.schedule(self.endService, self.stdMaintenanceDistribution())
        else:
            self.schedule(self.endService, self.brokenRepairDistribution())

    def beginBreakdownService(self):
        self.breakdownQLength -= 1
        self.numAvailableMechanics -= 1
        if (self.rngs[5].random() <= self.pStdRepair):
            self.schedule(self.endService, self.stdRepairDistribution())
        else:
            self.schedule(self.endService, self.brokenRepairDistribution())

    def endService(self):
        self.numAvailableMechanics += 1
        self.numAvailableVehicles += 1
        if (self.maintenanceQLength > 0):
            self.schedule(self.beginMaintenanceService, 0.0)
        elif (self.breakdownQLength > 0):
            self.schedule(self.beginBreakdownService, 0.0)
        breakdownTime = self.breakdownDistribution()
        if (breakdownTime <= self.maintenanceCycleInDays):
            self.schedule(self.breakdown, breakdownTime)
        else:
            self.schedule(self.maintenance, self.maintenanceCycleInDays)


    # init method kickstarts a simplekit model.  State variables are
    # set to initial values, and some preliminary events get scheduled
    def init(self):
        self.numAvailableVehicles = self.maxVehicles
        self.numAvailableMechanics = self.maxMaintainers
        self.maintenanceQLength = 0
        self.breakdownQLength = 0
        # self.numAvailableVehicles.times do |i|
        for _ in range(self.numAvailableVehicles):
            breakdownTime = self.breakdownDistribution()
            if (breakdownTime <= self.maintenanceCycleInDays):
                self.schedule(self.breakdown, breakdownTime)
            else:
                self.schedule(self.maintenance, self.maintenanceCycleInDays)
        self.schedule(self.shutDown, self.haltTime)
        # STDOUT.puts "DailyAvailabilityReport"
        self.schedule(self.dailyReport, 0.0)


if "__main__" == __name__:
    parser = ArgumentParser()
    parser.add_argument("maxVehicles", help="Initial Stock level", type=int)
    parser.add_argument("maxMaintainers", help="#Maintenance personnel", type=int)
    parser.add_argument("breakdownRate", help="Normal breakdown rate", type=float)
    parser.add_argument("maintenanceCycleDays", help="Maintenance cycle length (days)", type=int)
    parser.add_argument("pStdRepair", help="Proportion needing std repair", type=float)
    parser.add_argument("pStdMaint", help="Proportion receiving std maintenance", type=float)
    parser.add_argument("alpha", help="Alpha for Weibull Repair Distribution", type=float)
    parser.add_argument("beta", help="Beta for Weibull Repair Distribution", type=float)
    parser.add_argument("haltTimeInDays", help="Number of days to run", type=int)
    parser.add_argument("outputType", help="Output type: f=full, s=summary", choices=['f','s'])
    parser.add_argument("-r", "--rnseeds", nargs=6, default=[], help="Optional list of 6 seed values", type=int)
    args = parser.parse_args()

    # options = {}
    # OptionParser.new do |opts|
    #   opts.banner = "Usage: #{$PROGRAM_NAME} [-h|--help] [filenames...[]"
    #   opts.on("-r", "--rnseeds seed1,seed2,seed3,seed4,seed5,seed6", Array, "Provide six comma-separated integer random number seeds") do |rnseeds|
    #     # options[:rnseeds] = rnseeds.map(&:to_i) # Convert each element to Integer
    #    pp rnseeds.map
    #    pp rnseeds
    #     options[:rngs] = rnseeds.map do |seed|
    #       Xoroshiro::Random.new(seed: seed.to_i) # Convert each element to integer to seed rng
    #     end
    #   end
    # end.parse!

    # Check if we have exactly six integers

    # if options[:rngs].nil?:
    #     options[:rngs] = Array.new(6) { Xoroshiro::Random.new }
    # elif options[:rngs].size != 6:
    #     puts "Please provide six comma-separated integers using the -r option,"
    #     puts "or leave out the -r option to use system-generated seeds"
    #     sys.exit()

    # if ARGV.length != 10:
    #   err_msg
    # else:
    #   # Run model based on command-line arguments...
    #   maxVehicles = ARGV[0].to_i
    #   maxMaintainers = ARGV[1].to_i
    #   breakdownRate = ARGV[2].to_f
    #   maintenanceCycleInDays = ARGV[3].to_i
    #   pStdRepair = ARGV[4].to_f
    #   pStdMaint = ARGV[5].to_f
    #   # distributionChoice = ARGV[5].split('')[0].downcase
    #   alpha = ARGV[6].to_f
    #   beta = ARGV[7].to_f
    #   haltTimeInDays = ARGV[8].to_i
    #   outputType = ARGV[9].downcase

    if len(args.rnseeds) == 0:
        rngs = [np.random.default_rng() for _ in range(6)]
    else:
        rngs = [np.random.default_rng(seed=s) for s in args.rnseeds]

    # print(rngs)
    # exit()

    FleetAvail(
          args.maxVehicles,
          args.maxMaintainers,
          args.breakdownRate,
          args.maintenanceCycleDays,
          args.pStdRepair,
          args.pStdMaint,
          'w',
          args.alpha,
          args.beta,
          args.haltTimeInDays * 8.0,
          args.outputType,
          rngs
    ).run()

File: Models/FleetAvailability/fleet_avail_6seeds.rb
#!/usr/bin/env ruby

begin
  require 'simplekit'
  require 'quickstats'
  require 'random_variates'
  require 'optparse'
rescue LoadError => e
  missing = e.message.split[-1]
  STDERR.puts "\nERROR:  " + e.message
  STDERR.puts "\n\tRun the command \"gem install #{missing}\" to fix this."
  STDERR.puts "\t(This may require administrative privileges on Windows or"
  STDERR.puts "\tuse of the \"sudo\" command on OS X or linux.)"
  STDERR.puts
  exit
end


# Demonstration model of Fleet Availability

class FleetAvail
  include SimpleKit

  # model state
  attr_reader  :numAvailableVehicles,
    :numAvailableMechanics,
    :maintenanceQLength,
    :breakdownQLength

  # model parameters
  attr_reader  :maxVehicles,
    :maxMaintainers,
    :breakdownRate,
    :maintenanceCycleInDays,
    :repairRate,
    :haltTime,
    :outputType

  # the actual model implementation...

  # Constructor initializes the model parameters.
  def initialize(maxVehicles, maxMaintainers, breakdownRate,
                 maintenanceCycleInDays, pStdRepair,
                 pStdMaint, distributionChoice, alpha, beta, haltTime,
                 outputType, rngs)
    @maxVehicles = maxVehicles
    @maxMaintainers = maxMaintainers
    @breakdownRate = breakdownRate
    @maintenanceCycleInDays = maintenanceCycleInDays
    @pStdRepair = pStdRepair
    @pStdMaint = pStdMaint
    # @repairRate = 1.0 / meanRepairTime
    @distributionChoice = distributionChoice
    @haltTime = haltTime
    @outputType = outputType
    @rngs = rngs

    if (outputType == 's')
      STDOUT.printf "initial stock,"
      STDOUT.printf "#maintainers,"
      STDOUT.printf "breakdown rate,"
      STDOUT.printf "maintenance cycle,"
      # STDOUT.printf "repair time,"
      STDOUT.printf "p(std repair),"
      STDOUT.printf "p(std maint),"
      # STDOUT.printf "distribution,"
      STDOUT.printf "alpha,"
      STDOUT.printf "beta,"
      STDOUT.printf "run length"
      STDOUT.printf ",average availability,"
      STDOUT.printf "stddev availability,"
      STDOUT.printf "min,"
      STDOUT.printf "10th percentile,"
      STDOUT.printf "20th percentile,"
      STDOUT.printf "median\n"
      STDOUT.printf "%d,%d,%f,%d,%f,%f,%f,%f,%d", maxVehicles, maxMaintainers,
        breakdownRate, maintenanceCycleInDays, pStdRepair, pStdMaint,
        # distributionChoice,
        alpha, beta, haltTime / 8
    end

    if (outputType == 'f')
      STDOUT.printf "initial stock,"
      STDOUT.printf "#maintainers,"
      STDOUT.printf "breakdown rate,"
      STDOUT.printf "maintenance cycle,"
      STDOUT.printf "p(std repair),"
      STDOUT.printf "p(std maint),"
      STDOUT.printf "alpha,"
      STDOUT.printf "beta,"
      STDOUT.printf "run length"
      STDOUT.printf "%d,%d,%f,%d,%f,%f,%f,%f,%d", maxVehicles, maxMaintainers,
        breakdownRate, maintenanceCycleInDays, pStdRepair, pStdMaint,
        # distributionChoice,
        alpha, beta, haltTime / 8
      STDOUT.print "\n\nday,availability\n"
    end


    @breakdownDistribution = 
      RV::Exponential.new(rate: @breakdownRate, rng: rngs[0])

    @stdRepairDistribution = case @distributionChoice
    when 'e'
      RV::Exponential.new(rate: @repairRate, rng: rngs[1])
    when 'u'
      RV::Uniform.new(min: 0, max: 2.0 / @repairRate, rng: rngs[1])
    when 'w'
      RV::Weibull.new(rate: alpha, k: beta, rng: rngs[1])
    end

    @brokenRepairDistribution = case @distributionChoice
    when 'e'
      RV::Exponential.new(rate: @repairRate / 4.0, rng: rngs[2])
    when 'u'
      RV::Uniform.new(min: 0, max: 8.0 / @repairRate, rng: rngs[2])
    when 'w'
      RV::Weibull.new(rate: 4.0 * alpha, k: beta, rng: rngs[2])
    end

    @stdMaintenanceDistribution = RV::Uniform.new(min: 5.5/8, max: 6.5/8, rng: rngs[3])


  end

  # init method kickstarts a simplekit model.  State variables are
  # set to initial values, and some preliminary events get scheduled
  def init
    @dailyStats = QuickStats.new
    @results = []
    @numAvailableVehicles = @maxVehicles
    @numAvailableMechanics = @maxMaintainers
    @maintenanceQLength = 0
    @breakdownQLength = 0
    @numAvailableVehicles.times do |i|
      breakdownTime = @breakdownDistribution.next
      if (breakdownTime <= @maintenanceCycleInDays)
        schedule(:breakdown, breakdownTime)
      else
        schedule(:maintenance, @maintenanceCycleInDays)
      end
    end
    schedule(:shutDown, @haltTime)
    # STDOUT.puts "DailyAvailabilityReport"
    schedule(:dailyReport, 0.0)
  end

  # Event methods follow...

  def shutDown
    @results.sort!
    n = @results.length + 1
    if outputType == 's'
      STDOUT.printf ",%f,%f,%d,%d,%d,%d\n", @dailyStats.avg, @dailyStats.std_dev,
        @results[0], @results[0.1 * n], @results[0.2 * n], @results[0.5 * n]
    elsif outputType == 'f'
      STDOUT.printf "%d,%d\n", (model_time / 8.0).to_i, @numAvailableVehicles
    end
    halt
  end

  def dailyReport
    if model_time < @haltTime
      @results << @numAvailableVehicles
      @dailyStats.new_obs(@numAvailableVehicles)
      STDOUT.printf "%d,%d\n", (model_time / 8.0).to_i, @numAvailableVehicles if outputType == 'f'
    end
    schedule(:dailyReport, 8.0)
  end

  def breakdown
    @breakdownQLength += 1
    @numAvailableVehicles -= 1
    schedule(:beginBreakdownService, 0.0) if (@numAvailableMechanics > 0)
  end

  def maintenance
    @maintenanceQLength += 1
    @numAvailableVehicles -= 1
    schedule(:beginMaintenanceService, 0.0) if (@numAvailableMechanics > 0)
  end

  def beginMaintenanceService
    @maintenanceQLength -= 1
    @numAvailableMechanics -= 1
    if (@rngs[4].rand <= @pStdMaint)
      schedule(:endService, @stdMaintenanceDistribution.next)
    else
      schedule(:endService, @brokenRepairDistribution.next)
    end
  end

  def beginBreakdownService
    @breakdownQLength -= 1
    @numAvailableMechanics -= 1
    if (@rngs[5].rand <= @pStdRepair)
      schedule(:endService, @stdRepairDistribution.next)
    else
      schedule(:endService, @brokenRepairDistribution.next)
    end
  end

  def endService
    @numAvailableMechanics += 1
    @numAvailableVehicles += 1
    if (@maintenanceQLength > 0)
      schedule(:beginMaintenanceService, 0.0)
    else
      schedule(:beginBreakdownService, 0.0) if (@breakdownQLength > 0)
    end
    breakdownTime = @breakdownDistribution.next
    if (breakdownTime <= @maintenanceCycleInDays)
      schedule(:breakdown, breakdownTime)
    else
      schedule(:maintenance, @maintenanceCycleInDays)
    end
  end
end


def err_msg
  STDERR.puts
  warn "\nMust supply ten command-line arguments:\n"
  warn "\tInitial Stock level (int)"
  warn "\t#Maintenance personnel (int)"
  warn "\tNormal breakdown rate (double)"
  warn "\tMaintenance cycle length (int)"
  warn "\tProportion needing std repair"
  warn "\tProportion receiving std maintenance"
  # warn "\tMaintenance distribution choice [e|u|s|t]"
  warn "\tAlpha and Beta for Repair Distribution (double) (double)"
  warn "\tNumber of days to run (int)"
  warn "\tOutput type: f=full, s=summary (f/s)"
  # warn "\tOptionally, an integer can be provided for seeding the simulation"
  # warn "\nExample 1: ruby #{File.basename($0)} 50 2 0.01 90 0.8 0.95 0.2 1.5 50 s \n"
  # warn "\nExample 2: ruby #{File.basename($0)} 50 2 0.01 90 0.8 0.95 0.2 1.5 50 f 9876543\n"
  STDERR.puts
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: #{$PROGRAM_NAME} [-h|--help] [filenames...[]"
  opts.on("-r", "--rnseeds seed1,seed2,seed3,seed4,seed5,seed6", Array, "Provide six comma-separated integer random number seeds") do |rnseeds|
    # options[:rnseeds] = rnseeds.map(&:to_i) # Convert each element to Integer
   pp rnseeds.map
   pp rnseeds
    options[:rngs] = rnseeds.map do |seed|
      Xoroshiro::Random.new(seed: seed.to_i) # Convert each element to integer to seed rng
    end
  end
end.parse!

# Check if we have exactly six integers

if options[:rngs].nil?
  options[:rngs] = Array.new(6) { Xoroshiro::Random.new }
else
  if options[:rngs].size != 6
    puts "Please provide six comma-separated integers using the -r option,"
    puts "or leave out the -r option to use system-generated seeds"
    exit
  end
end

if ARGV.length != 10
  err_msg
else
  # Run model based on command-line arguments...
  maxVehicles = ARGV[0].to_i
  maxMaintainers = ARGV[1].to_i
  breakdownRate = ARGV[2].to_f
  maintenanceCycleInDays = ARGV[3].to_i
  pStdRepair = ARGV[4].to_f
  pStdMaint = ARGV[5].to_f
  # distributionChoice = ARGV[5].split('')[0].downcase
  alpha = ARGV[6].to_f
  beta = ARGV[7].to_f
  haltTimeInDays = ARGV[8].to_i
  outputType = ARGV[9].downcase
  FleetAvail.new(maxVehicles, maxMaintainers, breakdownRate, maintenanceCycleInDays,
                 pStdRepair, pStdMaint, 'w', alpha, beta, 8.0 * haltTimeInDays, outputType, options[:rngs]).run
end
